### PR TITLE
Fix prevent duplicate GPU usage in distributed processing

### LINKED ISSUE #3485: `PartialState` `split_between_processes` resulting in duplicated output with `padding=True`
### System Info

```Shell
accelerate==1.3.0
torch==2.2.2 and torch 2.4.0
python3.8 and 3.11
```

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

Run below script using `torchrun --nproc-per-node 2 script.py`

```python
from accelerate import PartialState
from accelerate.utils import gather_object

# Start up the distributed environment without needing the Accelerator.
distributed_state = PartialState()

prompts = [str(i) for i in range(4)]

batch_size = 2
tokenized_prompts = [prompts[i : i + batch_size] for i in range(0, len(prompts), batch_size)]

completions_per_process = []
with distributed_state.split_between_processes(tokenized_prompts, apply_padding=True) as batched_prompts:
    for batch in batched_prompts:
        generated_text = [f"{distributed_state.device}: {t}" for t in batch]
        completions_per_process.extend(generated_text)

completions_gather = gather_object(completions_per_process)
print(completions_gather)
# Drop duplicates produced by apply_padding in split_between_processes
completions = completions_gather[: len(prompts)]

```

I found that when the number of `prompts` is divisible by `batch_size`, I see `completions_gather` to be like this:
["cuda 0: 0", "cuda 0: 1", "cuda 0: 0", "cuda 0: 1", "cuda 1: 2", "cuda 1: 3", "cuda 1: 2", "cuda 1: 3"]

This suggested that GPU 0 returns `["cuda 0: 0", "cuda 0: 1", "cuda 0: 0", "cuda 0: 1"]` and GPU 1 returns `["cuda 1: 2", "cuda 1: 3", "cuda 1: 2", "cuda 1: 3"]`

---
By further looking at the source code, I found out that the issue occurred after this PR  https://github.com/huggingface/accelerate/pull/2781

https://github.com/huggingface/accelerate/blob/main/src/accelerate/state.py#L473

```python
if apply_padding:
    if isinstance(result, torch.Tensor):
        from accelerate.utils import pad_across_processes, send_to_device

        # The tensor needs to be on the device before we can pad it
        tensorized_result = send_to_device(result, self.device)
        result = pad_across_processes(tensorized_result, pad_index=inputs[-1])
    else:
        result += [result[-1]] * (num_samples_per_process + 1 - len(result))
return result
```

If my input is `[[0,1], [2,3]]`, `num_samples_per_process` equals 1 because I have 2 batches and 2 GPUs.

On GPU 0, the item assigned to it is [[0,1]]. `(num_samples_per_process + 1 - len(result))` is `1+1-1=1`. The final result is therefore `2 * [[0,1]] = [[0,1], [0,1]]` which created the duplicates

It will work correctly if my input is `[[0,1], [2,3], [4]]`. 
- In this case GPU 0 gets [[0,1], [2,3]], `(num_samples_per_process + 1 - len(result))` is `1+1-2=0`. It doesn't create any duplicates.
- GPU 1 gets [[4]], and `(num_samples_per_process + 1 - len(result))` is `1+1-1=1`.  Therefore it got padded to [[4], [4]].

Maybe this line can be changed to 
`result += [result[-1]] * (num_samples_per_process + (1 if num_extras>0 else 0) - len(result))`

### Expected behavior

`[{"cuda 0: 0", "cuda 0: 1", "cuda 1: 2", "cuda 1: 3"]`

### PR DESCRIPTION
**Pull request by ved1beta**
# What does this PR do?

This PR fixes the issue where running distributed processing with more processes than available GPUs would cause NCCL errors due to duplicate GPU usage. The fix includes proper CUDA device initialization and validation in the `PartialState` class.

## Key Changes
- Add proper CUDA device initialization in `PartialState`
- Add device count validation before distributed processing
- Add warning when attempting to use more processes than available GPUs
- Fix NCCL error when multiple processes try to use same GPU

## Test Code
```python
from accelerate import PartialState
from accelerate.utils import gather_object
import os
import torch

# Print CUDA information
print(f"CUDA available: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"Number of CUDA devices: {torch.cuda.device_count()}")
    print(f"Current CUDA device: {torch.cuda.current_device()}")
    print(f"CUDA device name: {torch.cuda.get_device_name()}")

# Initialize distributed state first
distributed_state = PartialState()

# Create some test data
prompts = [str(i) for i in range(4)]  # ["0", "1", "2", "3"]
batch_size = 2

# Split into batches
tokenized_prompts = [prompts[i : i + batch_size] for i in range(0, len(prompts), batch_size)]

completions_per_process = []
with distributed_state.split_between_processes(tokenized_prompts, apply_padding=True) as batched_prompts:
    for batch in batched_prompts:
        generated_text = [f"{distributed_state.device}: {t}" for t in batch]
        completions_per_process.extend(generated_text)

# Gather results from all processes
completions_gather = gather_object(completions_per_process)
print(completions_gather)
```

## How to Test
1. Single GPU test:
```bash
CUDA_VISIBLE_DEVICES=0,1 torchrun --nproc-per-node 1 test.py
```

2. CPU mode test (for distributed testing):
```bash
CUDA_VISIBLE_DEVICES="" torchrun --nproc-per-node 2 test.py
```

## Expected Output
For single GPU test:

CUDA available: True
Number of CUDA devices: 1
Current CUDA device: 0
CUDA device name: NVIDIA GeForce RTX 3050 6GB Laptop GPU
Local rank: 0
Using device: cuda:0
['cuda:0: 0', 'cuda:0: 1', 'cuda:0: 2', 'cuda:0: 3']

Fixes #3485

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

@SunMarc @BenjaminBossan @zach-huggingface - This PR involves core parts of the library and distributed processing functionality.

### CODE REVIEW

**Review by SunMarc — COMMENTED** [review]
Thanks ! Please add a test for this specific case and don't forget to credit the other PR !

File: tests/test_partial_state_device.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            # This should execute without raising NCCL errors due to duplicate GPU usage
+            result = execute_subprocess_async(cmd, env=os.environ.copy())
+            self.assertEqual(result.returncode, 0, f"Process failed with error: {result.stderr.decode('utf-8')}")
+    
+    @require_cuda
+    def test_device_assignment_warning(self):
+        """
+        Tests that a warning is raised when PartialState is initialized with more processes 
+        than available GPUs.
+        """
```
Comment (SunMarc) [inline tests/test_partial_state_device.py:68]: not sure what you are trying to do here, this is not the goal of this pr

**Review by SunMarc — COMMENTED** [review]
Please update the existing tests for test_split_between_processes_list instead of creating new ones

File: tests/test_partial_state_device.py (hunk 1 above)
Comment (ved1beta) [reply tests/test_partial_state_device.py:68]: removing these files , adding test to `test_split_between_processes_list`
